### ISSUE TITLE
Adding support for Arm64

### ISSUE THREAD

**Issue by gilesknap** [epics-containers/epics-base#12]
Intro
------

I have successfully got arm64 builds of epics-base to work.

However there are a couple of reasons not to merge this into main. 

- DLS does not actually need arm64
- arm64 builds are prohibitively slow on Github Actions - see more details below.

Arm at DLS
---------------

Arm64 is useful for people using new macs for dev or running IOCs.

DLS's arm IOCs run on 32bit architectures armv7 and armv8 both of which have cross compilation in our current build system. 

Even if we wanted to switch to doing native compile of these IOCs we could not as the 32 bit arm variants are lacking in package support and are not supported on GHA.

Hence these would be supported in containers as cross compile on linux:x86_64 just like RTEMS.

Arm64 Build Performance
----------------------------------

I have tried two mechanisms for targeting arm64. 

- use buildx and specify platform=linux/arm64,linux/amd64
- do a native MacOS build using the os matrix in GHA

In both cases, this snippet in the Dockerfile instructs the epics build system of the target os:
```Dockerfile
FROM base AS environment-amd64
ENV EPICS_HOST_ARCH=linux-x86_64

FROM base AS environment-arm64
ENV EPICS_HOST_ARCH=linux-arm

FROM environment-$TARGETARCH AS environment
```

The buildx approach uses QEMU to emulate arm64 for the compilation. This results in a build time of 1hr18m, see https://github.com/epics-containers/epics-base/actions/runs/5155081287

I had hoped that a native build on MacOS would improve things, but for whatever reason this is way worse at 3hr36m https://github.com/epics-containers/epics-base/actions/runs/5155705777.

**Comment by gilesknap** [comment]
@coretl @ulrikpedersen tagging you both for a discussion of how to proceed with this.

It is possible that my native arm64 compile was done wrong in some fashion as it seems a very bad result.

**Comment by gilesknap** [comment]
I was  wondering if the cause of the slow native build is to do with target.
We are targetting linux:arm as that is what will be inside the ubuntu container running on a Mac.
But the native compiler for Mac would be darwin-aarch64.

Would this maybe cause the epics build system to use cross compilation? The answer should be no as the build itself is running inside a linux container. Looking at the logs is looks to be targeting linux:arm and using gcc - I don't see any cross compilation.

**Comment by gilesknap** [comment]
It occurs to me that even if GitHub has real Macs for GHA they are probably not M1 as that is less than 3 years old. We are probably running on a x86 mac and getting QEMU to arm64 emulate for us!

doh!

**Comment by gilesknap** [comment]
@ulrikpedersen I'm merging into main all the changes except with arm64 approaches commented out. I can confirm that M1 is coming to GHA and will enable the native compile when this happens. 

See https://github.com/github/roadmap/issues/528

The two approaches to building for arm64 are commented in .github/workflows/build.yml, see snippet below. The first approach is the fastest for now with build time 1h20m.

https://github.com/epics-containers/epics-base/blob/96f1abad791614caa03b8cd9bae22b36f083d79b/.github/workflows/buiild.yml#L7-L41

**Comment by ulrikpedersen** [comment]
> I can confirm that M1 is coming to GHA and will enable the native compile when this happens.

Very cool 👍 
I was considering offering up my M1 Mini at home as a self-hosted runner. I even spent a couple of hours trying to figure out how - but I couldn't work out how to do it in a very safe way. It is not a dedicated machine for this task and has all my personal data on it.

In other news: I have bought a PC with Intel CPU and installed Ubuntu on it so that I can run both epics-containers and Tango containers without all the faff 😄 
Looking forward to the day when GH releases Mac M1 runners though 🤞

**Comment by gilesknap** [comment]
Yeah self hosted runners are best done on isolated machines. We are planning to use machines in the IRIS cloud to do it.

FYI: as an esteemed customer of epics-containers I should warn you that Tom Gary and I have just sketched out a major refactor. The good news is that it should simplify the developer experience and align more closely with the standard devcontainer workflow. The bad news is its a bit of a rewrite of all the tutorials and I wont get started on that until mid Aug.

**Comment by ulrikpedersen** [comment]
Thanks for the heads-up. I don't think it will affect us in how we are currently using epics-containers because we haven't fully adopted the workflow outlined in the tutorials...

**Comment by gilesknap** [comment]
@ulrikpedersen now is a good time to adopt the full epics-containers workflow if you were considering it.

- Two major re-writes since you last used it and it is now looking stable.
- all the tutorials are up to date with the latest changes
- generic ioc and beamline repositories are created using a copier template so you can easily adopt any future changes
- I've just finished re-doing RTEMS and would be happy to take another look at arm64 if you still have an interest

**Comment by amichelotti** [comment]
Hello Giles, all. We are using your images and I think providing an ARM64-based developer image would be very useful, as x86-based Macs are becoming less common and servers with ARM-based CPUs are increasingly popular.

**Comment by gilesknap** [comment]
@amichelotti does that mean you need linux-arm64 and darwin-arm64 which is different because apple kernel is not linux.

I once had linux-arm64 working and this PR is attempting to re-instate that, but I've not had time to verify it yet https://github.com/epics-containers/epics-base/pull/50/changes 

The intention is that all supported linux flavours would be provided by a single multi-arch container image. However there is a little work to do to get that to happen.

**Comment by gilesknap** [comment]
Status update and a suggested way forward, since the main blocker in the original report no longer applies:

- **Build speed**: GitHub now has native arm64 Linux runners (`ubuntu-24.04-arm`), so builds no longer run under QEMU, which took 1h18m. The open PR #50 uses them, and both its `linux-x86_64` and `linux-aarch64` jobs passed.
- **Base image**: #50 switches the arm job to `ubuntu:24.04` and so has to add `build-essential` and copy in `uv`. That should not be needed: `ghcr.io/diamondlightsource/ubuntu-devcontainer:noble`, the default `BASE_IMAGE`, is already a multi-arch index with `linux/amd64` and `linux/arm64` entries. I checked the manifest on ghcr. If the arm job keeps the default base image, the Dockerfile can stay unchanged.
- **Naming**: #50 publishes a separate `epics-base-aarch64-*` image. The conclusion in #36 was one multi-arch `epics-base-developer`/`epics-base-runtime` for native builds, with only cross-compile targets named separately. That fits the "multi-runner" option from #36: build each platform on its own native runner and push by digest, then have a final job run `docker buildx imagetools create` to merge the digests under the tag. The objection in #36 was that x86_64 would have to wait for the slow arm build. With native runners the arm build takes roughly as long as x86_64, so that objection is mostly gone.
- **Still open**: generic IOCs would also need arm64 builds (ioc-template CI and any ibek-support recipes that hard-code `x86_64`, e.g. `PCRE_LIB=/usr/lib/x86_64-linux-gnu` in StreamDevice). @amichelotti, if you only need a developer image on arm64 machines, the epics-base part above is enough to start with.

_Triage comment drafted by Claude Code._
